Name the sections by strings, and add a size for all sections.

JSStats opened the pull request.

jfbastien commented on BinaryEncoding.md:
```diff
@@ -73,27 +73,37 @@ Strings referenced by the module (i.e. function names) are encoded as null-termi
+| Field | Type | Description |
+| ----- |  ----- | ----- |
+| id_len | `varuint32` | section identifier string length |
+| id_str | `char*` | section identifier string |
```
In general we don't want `char*`, we instead want a leb128 size followed by a byte string (it doesn't even have to be valid Unicode).

jfbastien commented on BinaryEncoding.md:
```diff
@@ -73,27 +73,37 @@ Strings referenced by the module (i.e. function names) are encoded as null-termi
 
 ### Memory section
+
+ID: `memory.wasm`
```
I'd drop the `.wasm` from all the strings.

titzer replied:
Can we drop the .wasm? It's cleaner.

titzer commented on BinaryEncoding.md:
```diff
@@ -73,27 +73,37 @@ Strings referenced by the module (i.e. function names) are encoded as null-termi
 ## High-level structure
-A module contains a sequence of sections, each identified by a *section identifier byte* whose encodings
-are listed below.
+A module contains a sequence of sections. Sections can in general be repeated, but some can occur only once or have dependant sections that must preceed them. Each section is identified by an immediate string. Sections whose identity is unknown to the WebAssembly implementation are ignored and this is supported by including the size in bytes for all sections. The encoding of all sections begins a ...
```
How about a fixed order for sections? That means checking for duplicates is easy and dependencies are always correct by construction. It also means that an engine doesn't need to do a string-table lookup to decide how to handle the next section; it always has a next expected section and can skip unexpected sections in between.

jfbastien replied:
Except we need to allow unknown sections in the middle of that fixed order.

JSStats replied:
It would seem a significant and orthogonal change to give all the wasm sections a fixed order. It seems unnecessary from a source code point of view and might be frustrating even for the wast if it too required a fixed order, but can be worked by the tooling and in meta information, so if people could just confirm this decision I'm happy to note this here. But could this be split out into a follow up PR that someone else could champion? Or are you concerned to avoid the complexity of requiring a table lookup and want to settle this now?

jfbastien replied:
Yes, fixed order is something we want, but I agree it's something we can do in a separate PR.

jfbastien commented on BinaryEncoding.md:
```diff
@@ -205,16 +227,6 @@ stored to from dedicated instructions.
-| size  | `varuint32` | size of this section in bytes | 
-| body  | `bytes`     | contents of this section |
-
-Sections whose ID is unknown to the WebAssembly implementation are ignored.
```
The intent is to still mention unknown sections as ignored, but specify no content (as the high-level structure is the same).

titzer commented on BinaryEncoding.md:
```diff
@@ -73,27 +73,37 @@ Strings referenced by the module (i.e. function names) are encoded as null-termi
+| ----- |  ----- | ----- |
+| id_len | `varuint32` | section identifier string length |
+| id_str | `char*` | section identifier string |
+| size  | `varuint32` | size of this section in bytes, excluding the ID string and this size |
```
How about putting the size first, which makes it easier to simply stride through the sections.

titzer commented on BinaryEncoding.md:
```diff
@@ -166,31 +182,37 @@ A module may only contain one data segments section.
 
 ### Indirect Function Table section
+
+ID: `table.wasm`
```
idft or indirect_table?

jfbastien replied:
We can uses spaces too! "indirect table" :smile:

jfbastien commented:
@JSStats can you please update with the names in this patch: https://codereview.chromium.org/1743773002/

JSStats commented:
Thank you. Updated the section names and addressed much of the other feedback:

* Used the v8 patch names. Except the start_function section which seem todo.

* Moved the size first, before the name, to allow walking the section without reading the names.

* Clarified that the unknown sections can occur in any order and between wasm sections that have an ordering constraint.

* Explicitly mentioned unknown sections.

There remains the matter of a decision to use a fixed order for all wasm sections, but perhaps this could be handled in a separate follow up issue.

ptal

jfbastien commented:
lgtm

titzer commented:
lgtm

On Fri, Feb 26, 2016 at 5:20 PM, JF Bastien <notifications@github.com>
wrote:

> lgtm
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WebAssembly/design/pull/557#issuecomment-189546216>.
>
